[August] 미션4: 모든 기물 배치하기

polynomeer opened the pull request:
### 미션4: 모든 기물 배치하기

안녕하세요! 좋은 리뷰 항상 감사합니다. '미션4: 모든 기물 배치하기'를 구현하고 주어진 테스트 코드가 잘 실행되는 것을 확인했습니다. 구현 내용은 아래와 같습니다.

### 구현 내용

#### main/chess

- Board.java

  - 기존에는 `whitePieces`와 `blackPieces` 리스트에 `Pawn`을 추가했었는데, 다른 체스말들을 따로 추가하기 위해서 `whitePawns`와 `blackPawns` 리스트를 별도로 분리했습니다.
  - 모든 add메서드에서 `color`를 검증하는 코드가 있었는데, 이 부분을 `Piece::isWhtie()` 와 `Piece::isBlack()`으로 대체하였습니다.
  - Game에 있던 `print()`메서드를 Board의 `toString()`에서 String을 반환하는 형태로 변경했습니다. 이렇게 하면 Board를 나타내는 String으로서 `toString()`의 쓰임새와 맞기도 하고, Board클래스에 속해있는게 적절하지 않을까 생각했습니다.
  
#### main/pieces

기존에 있던 `Pawn.java`를 삭제했습니다. `Piece`를 모든 체스말들이 상속받아서 각 클래스로 존재하는 것을 생각하고 설계한 것인데, 생각보다 그 체스말이 각 클래스로 구현될 정도의 필요가 없다고 생각해서 미션의 요구사항에 맞게 `Piece`가 모든 체스말들의 속성을 갖고 있는 형태로 변경했습니다.

- Piece.java
  - 테스트 코드를 미리 작성하고, 각 체스말의 표기법을 상수 필드로 선언했습니다.
  - `color`과 `representation` 필드를 가지며, 매개변수로 두개의 속성을 모두 전달받는 private 생성자만 남겨두어서 외부에서 직접 생성하지 못하도록 했습니다.
  - 각 체스말을 생성하는 각 create 메서드에서 생성자를 호출합니다.
  - `isWhite()`와 `isBlack()`은 체스말의 `color`를 검증합니다.

#### main/utils

- StringUtils.java
  - 유틸리티 클래스로서 외부에서의 인스턴스화를 막기위해 private 생성자를 가집니다.
  - appendNewLine(String str) : 전달받은 str에 개행문자를 추가해서 리턴합니다.

#### test/chess

- BoardTest.java
  - `Board::size()`으로 제대로 체스말이 생성되었는지 개수를 검증합니다.
  - `Board::toString()`을 통해 전체 체스판의 모양이 제대로 출력되는지 검증합니다. 

- PieceTest.java
  - `verifyColor()`에서 `isWhite()`와 `isBlack()`에 대한 검증을 수행합니다.

polynomeer committed:
feat: Add StringUtils in utils package

Added utils package and StringUtils class in it.
Added appendNewLine() method in StringUtils class,
and Defined as static to use for utility.
Defined private constructor to prevent from getting instance of StringUtils class.

polynomeer committed:
refactor: Move Game::print to Board::toString

Moved Game::print() method to Board::toString() method.
toString return the state of board including white and black pieces.
To use StringUtils.appendNewLine() method, declared static import.

polynomeer committed:
refactor: Change Pawn to Piece

Deleted abstract Piece class.
Changed pieces.Pawn class to pieces.Piece .
All methods using 'Pawn' changed to use 'Piece'.
Removed unnecessary throws exceptions.

polynomeer committed:
feat: Add factory methods to create all pieces

Deleted BoardTest and PawnTest.
Added PieceTest and test codes to add create pieces methods.
Added the factory methods which create white and black pieces by TDD.

polynomeer committed:
feat: Add BoardTest to test Board

Added BoardTest to test Board class.
It tests Board::pieceCount() and Board::toString().

polynomeer committed:
feat: Add isWhite and isBlack methods

Added isWhite() and isBlack() method in Piece class
to check the color of piece is correctly assigned.
They are defined by the test code in PieceTest.

polynomeer committed:
refactor: Change verifying logic to use isWhite and isBlack

Changed condition in verifying color logic to use isWhite() and isBlack() methods in Board class.

polynomeer committed:
feat: Add separated List of pawns

Added whitePawns and blackPawns List to handle separately.
Added addWhitePawns() and addBlackPawns() to add pawns to separated list.
This codes pass the BoardTest code.

sonypark commented on src/main/java/utils/StringUtils.java:
```diff
@@ -0,0 +1,12 @@
+package utils;
+
+public class StringUtils {
+    public static final String NEWLINE = System.getProperty("line.separator");
+
```
`NEWLINE` 은 내부에서만 사용하는 것 같은데, 접근제어자는 private으로 바꿔주세요~

polynomeer replied:
네, 수정사항 반영했습니다!

sonypark commented on src/main/java/chess/Board.java:
```diff
@@ -1,30 +1,50 @@
 
+import static utils.StringUtils.appendNewLine;
+
 public class Board {
     public static final int BOARD_SIZE = 8;
```
BOARD_SIZE 도 private으로 둬도 될 것 같네요

polynomeer replied:
네, 반영했습니다!

sonypark commented on src/test/java/chess/BoardTest.java:
```diff
@@ -3,67 +3,29 @@
-    @DisplayName("Init board")
-    public void initialize() throws Exception {
+    @DisplayName("Verify Board initialize and toString")
+    public void create() throws Exception {
```
`throws Exception` 은 왜 필요할까요~~?

polynomeer replied:
필요가 없어서 제거했습니다!

sonypark commented on src/main/java/pieces/Piece.java:
```diff
@@ -1,22 +1,90 @@
     private final String color;
     private final char representation;
 
-    public Piece(String color, char representation) {
+    private Piece(String color, char representation) {
         this.color = color;
         this.representation = representation;
     }
 
+    public static Piece createWhitePawn() {
+        return new Piece(WHITE_COLOR, WHITE_PAWN_REPRESENTATION);
+    }
+
+    public static Piece createBlackPawn() {
+        return new Piece(BLACK_COLOR, BLACK_PAWN_REPRESENTATION);
+    }
+
+    public static Piece createWhiteKnight() {
+        return new Piece(WHITE_COLOR, WHITE_KNIGHT_REPRESENTATION);
+    }
+
+    public static Piece createBlackKnight() {
+        return new Piece(BLACK_COLOR, BLACK_KNIGHT_REPRESENTATION);
+    }
```
👍

sonypark commented on src/main/java/chess/Game.java:
```diff
@@ -24,7 +24,7 @@ static void process(Scanner scanner) {
             if (command.equals("start")) {
                 Board board = new Board();
                 board.initialize();
-                print(board);
+                System.out.println(board);
             } else if (command.equals("end")) {
```
게임시작과 종료를 의미하는 start, end는 자바 상수로 둬도 될 것 같아요:)

polynomeer replied:
우선은 Game의 private 상수 필드로 두었습니다.

sonypark commented on src/main/java/chess/Board.java:
```diff
@@ -1,30 +1,50 @@
     public void addBlack(Piece piece) {
-        if (piece.getColor().equals(Piece.BLACK_COLOR)) {
+        if (piece.isBlack()) {
             blackPieces.add(piece);
         }
     }
 
+    public void addWhitePawn(Piece piece) {
+        if (piece.isWhite()) {
+            whitePawns.add(piece);
+        }
+    }
+
+    public void addBlackPawn(Piece piece) {
+        if (piece.isBlack()) {
+            blackPawns.add(piece);
+        }
+    }
+
     public int size() {
-        return whitePieces.size() + blackPieces.size();
+        return whitePieces.size()
+                + blackPieces.size()
+                + whitePawns.size()
+                + blackPawns.size();
```
`size()`가 체스판 크기를 리턴하는 메서드라면 위 사이즈가 체스판의 사이즈가 맞을까요?

polynomeer replied:
모호한 의미를 보다 명확히 하기위해 `sizeofPiece()`로 변경해보았습니다.

sonypark commented on src/main/java/chess/Board.java:
```diff
@@ -1,30 +1,50 @@
+
     private final List<Piece> whitePieces = new ArrayList<>();
     private final List<Piece> blackPieces = new ArrayList<>();
+    private final List<Piece> whitePawns = new ArrayList<>();
+    private final List<Piece> blackPawns = new ArrayList<>();
 
     public void addWhite(Piece piece) {
-        if (piece.getColor().equals(Piece.WHITE_COLOR)) {
+        if (piece.isWhite()) {
             whitePieces.add(piece);
         }
     }
 
     public void addBlack(Piece piece) {
-        if (piece.getColor().equals(Piece.BLACK_COLOR)) {
+        if (piece.isBlack()) {
             blackPieces.add(piece);
         }
     }
 
+    public void addWhitePawn(Piece piece) {
+        if (piece.isWhite()) {
+            whitePawns.add(piece);
+        }
+    }
+
+    public void addBlackPawn(Piece piece) {
+        if (piece.isBlack()) {
+            blackPawns.add(piece);
+        }
+    }
```
메서드 이름만 보면 addWhitePawn, addBlackPawn은 특정 color의 Pawn만 받을 것 같은데 내부에서는 color만 체크해주고 있네요🤔

polynomeer replied:
Pawn만 받도록 검증을 하는 코드를 Piece클래스에 별도로 추가하였습니다. Piece 내부에 `isWhitePawn()`과 `isBlackPawn()`으로 추가하여 Board에서 직접 Piece.representation 필드에 직접 접근하지 못하도록 했습니다.

sonypark commented on src/main/java/pieces/Piece.java:
```diff
@@ -1,22 +1,90 @@
     public static final String WHITE_COLOR = "white";
     public static final String BLACK_COLOR = "black";
+    public static final char WHITE_PAWN_REPRESENTATION = 'p';
+    public static final char BLACK_PAWN_REPRESENTATION = 'P';
+    public static final char WHITE_KNIGHT_REPRESENTATION = 'n';
+    public static final char BLACK_KNIGHT_REPRESENTATION = 'N';
+    public static final char WHITE_ROOK_REPRESENTATION = 'r';
+    public static final char BLACK_ROOK_REPRESENTATION = 'R';
+    public static final char WHITE_BISHOP_REPRESENTATION = 'b';
+    public static final char BLACK_BISHOP_REPRESENTATION = 'B';
+    public static final char WHITE_QUEEN_REPRESENTATION = 'q';
+    public static final char BLACK_QUEEN_REPRESENTATION = 'Q';
+    public static final char WHITE_KING_REPRESENTATION = 'k';
+    public static final char BLACK_KING_REPRESENTATION = 'K';
 
     private final String color;
     private final char representation;
 
-    public Piece(String color, char representation) {
+    private Piece(String color, char representation) {
         this.color = color;
         this.representation = representation;
     }
```
이렇게 representation을 표시하기 위한 상수가 많아지면서 이걸 어떻게 관리하면 좋을지 고민을 하시게 되겠네요!

polynomeer replied:
이 부분은 다음 미션에서 진행해보겠습니다! 리뷰 감사합니다! 😀

sonypark reviewed (commented):
안녕하세요 August, 리뷰어 소니 입니다.😊
요구사항대로 깔끔하게 구현 잘 하셨네요!💯
리뷰를 하면서 크게 고쳐야할 부분은 없었습니다.
생각해볼 문제에 대해 간단히 코멘트 남겼으니 확인해보시고 다음 단계로 넘어가시면 될 것 같아요.
수고하셨습니다👍

polynomeer committed:
chore: Delete unnecessary throws Exception

Deleted unnecessary throws Exception code.
Changed assertTrue() to assertThat().
Changed codes to assertThat(actual).isEqualTo(expected) format.

polynomeer committed:
chore: Delete unnecessary Main class

Deleted unnecessary Main class to simplify code.

polynomeer committed:
refactor: Change public fields to private

Changed some public fields to private fields
to do information hiding.

polynomeer committed:
refactor: Add final constant fields START, END

Added final constant fields START, END to clarify code and prevent from misusing.

polynomeer committed:
refactor: Change size() to sizeofPiece()

Changed the method name 'size()' to 'sizeofPiece()'
to clarify the meaning of piece.

polynomeer committed:
feat: Add isWhitePawn and isBlackPawn method

Added isWhitePawn() and isBlackPawn() methods in Piece class to achieve encapsulation.
Added test codes in PieceTest to verify isWhitePawn() and isBlackPawn().

sonypark reviewed (approved):
수고하셨습니다.
다음 단계 진행해주세요~
